redis/mutation_utils.cc

```diff
@@ -78,11 +102,63 @@ mutation make_mutation(service::storage_proxy& proxy, const redis_options& optio
     return m;
 }
 
-future<> write_strings(service::storage_proxy& proxy, redis::redis_options& options, bytes&& key, bytes&& data, long ttl, service_permit permit) {
+future<bool> write_strings(service::storage_proxy& proxy, redis::redis_options& options, bytes&& key, bytes&& data, long ttl, service_permit permit, smp_service_group& ssg) {
+    // construct cas_request
     db::timeout_clock::time_point timeout = db::timeout_clock::now() + options.get_write_timeout();
-    auto m = make_mutation(proxy, options, std::move(key), std::move(data), ttl);
+    auto schema = get_schema(proxy, options.get_keyspace_name(), redis::STRINGs);
+    const column_definition& column = *schema->get_column_definition(redis::DATA_COLUMN_NAME);
+    auto pkey = partition_key::from_single_value(*schema, key);
+    auto dk = dht::decorate_key(*schema, std::move(pkey));
+    auto partition_range = dht::partition_range::make_singular(dk);
+    dht::partition_range_vector partition_ranges;
+    partition_ranges.emplace_back(std::move(partition_range));
     auto write_consistency_level = options.get_write_consistency_level();
-    return proxy.mutate(std::vector<mutation> {std::move(m)}, write_consistency_level, timeout, nullptr, permit);
+    auto request = seastar::make_shared<write_strings_cas_request>(proxy, options, key, data, ttl);
```

Comment:
I think the use-after-free bug I worried about earlier indeed materializes here... key and data are local variables, but you save references to them in "request". When this function returns a future, these references will become dead.
I think you should move the keys *into* the write_strings_cas_request object. If you need them after the movement, you can always add an accessor method to this object, to read the key it holds.

Reply:
Stopped to pass `key` and `data` as references, switched to use std::move().